### PR TITLE
Sandbox recordset websocket

### PR DESCRIPTION
**Pull request by hatched**
When the GUI is deployed in sandbox mode it still needs to be able to process bundles. This adds websocket support to the getChangeSet environment endpoints.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2525/
Test PASSed.

**Comment by hatched** [comment]
#### To QA

```
mkdir -p ~/charms/trusty
cd ~/charms/trusty 
bzr branch lp:~juju-gui/charms/trusty/juju-gui/trunk juju-gui
juju bootstrap local
juju deploy --repository=~/charms local:trusty/juju-gui
juju set juju-gui juju-gui-source="https://github.com/hatched/juju-gui.git sandbox-recordset-websocket"
juju set juju-gui sandbox=true
```

Once it's up and running you should be in sandbox mode.
Drag a bundle to the canvas and it should generate a set of uncommited items on the canvas.
Commit these changes and it should deploy like a real environment.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2531/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2533/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2538/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2540/
Test FAILed.

**Comment by mitechie** [comment]
Tests passed on your other branch so I'm not sure this is spurious. That passing test was 13hrs ago and in between your first and latest failures here.

File: app/app.js [hunk 1]
```diff
@@ -456,8 +456,12 @@ YUI.add('juju-gui', function(Y) {
             credentials['user-' + envOptions.user] = envOptions.password;
             state.set('authorizedUsers', credentials);
           }
-          envOptions.conn = new sandboxModule.ClientConnection(
-              {juju: new sandboxModule.GoJujuAPI({state: state})});
+          envOptions.conn = new sandboxModule.ClientConnection({
+            juju: new sandboxModule.GoJujuAPI({
+              state: state,
+              socket_url: envOptions.socket_url
```
Comment (mitechie) [inline app/app.js]: oh interesting. We now require a socket here. How does this work in development?

**Comment by mitechie** [comment]
I'm curious how this works in development when you don't have a charm handy. Should this be like the charmstore api urls and be something we talk to the demo.jujucharms.com production instance to by default unless it's a live env deployment?

File: app/app.js (hunk 1 above)
Comment (hatched) [inline app/app.js]: It has an invalid url and any request tries to complete 3 times then fails and throws a notification to the user.

**Comment by hatched** [comment]
That's a good idea. We'll have to set a proper default socket url in the config (which was actually removed somewhat recently)

**Comment by hatched** [comment]
According to @frankban the charm does not support listening to websockets when in sandbox mode at the moment. So we'll have to postpone this functionality until a later date. The GUI supports dropping a  recordSet on the canvas to import a bundle which could have been generated from the bundlelib cli.

**Comment by mitechie** [comment]
Cool yea I think that'd be a nice way to go

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2541/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2542/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2543/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2544/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2546/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2547/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2548/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2549/
Test FAILed.

**Comment by makyo** [comment]
QA generally okay, but can you check on something for me? I deployed a bundle, commmitted the changes, then tried deleting the services, but was stopped by console errors.  They mostly have to do with scheduling unit deletion, which doesn't seem to come into play in this branch, but if you get a chance, can you try and reproduce? Specifically, leaving units unplaced

**Comment by makyo** [comment]
Sorry, generally :+1: because I suspect the problem is related to what I'm working on.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2550/
Test PASSed.

**Comment by hatched** [comment]
Thanks for the reviews! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci.jujugui.org:8080/job/juju-gui-merge/892

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**